### PR TITLE
gh-123290: fix reference leaks in `curses`

### PR DESCRIPTION
**Pull request by picnixz**
This will superseed https://github.com/python/cpython/pull/123291. With this PR, the curses module does not leak anymore :partying_face: 

* Issue: gh-123290

### CODE REVIEW

File: Modules/_cursesmodule.c [hunk 1]
```diff
@@ -161,12 +161,35 @@ typedef chtype attr_t;           /* No attr_t type is available */
...
+
+static inline _cursesmodule_state *
+get_cursesmodule_state_by_cls(PyTypeObject *cls)
+{
+    void *state = PyType_GetModuleState(cls);
+    assert(state != NULL);
+    return (_cursesmodule_state *)state;
+}
+
+// A global borrowed reference to the type error is needed for
```
Comment (picnixz) [inline Modules/_cursesmodule.c]:
```suggestion
// A global borrowed reference to the exception is needed for
```

**Comment by picnixz** [comment]
Oh, before I forget, I think clinic does not allow me to do something like

```C
module.cls.meth

	cls: defining_class
	[
		some_optional
	]
	/
```

I'll try to confirm this tomorrow (I don't know whether it's a bug or not)

File: Misc/NEWS.d/next/Library/2024-09-03-13-51-15.gh-issue-123290.IDR-iN.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fix reference leaks for the :mod:`curses` and :mod:`curses.panel` modules.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-09-03-13-51-15.gh-issue-123290.IDR-iN.rst]:
```suggestion
Fix reference leaks for the :mod:`curses` module.
```

**Comment by vstinner** [comment]
@erlend-aasland @encukou @ericsnowcurrently: This PR converts the `_curses` extension to multiphase initialization API (PEP 489) and convert static types to heap types. Do you recall if we decided to leave the `_curses` extension unchanged on purpose? Or is it ok to have multiple instances of this extension?

**Comment by picnixz** [comment]
By the way, with this PR and #123285, no modules should leak references (at least on POSIX).

**Comment by encukou** [comment]
> Do you recall if we decided to leave the `_curses` extension unchanged on purpose? 

I guess it's not very useful outside the main interpreter, so it was left behind.

I haven't looked at the code in detail, but, `_curses` manages process-global state (*the* terminal & the curses API around it), so it should:
- ideally: maintain process-global state, with all Python-level objects being “proxies” to that (i.e.  module-level state)
- or: only be loadable *once* per process

But neither had priority, so it's left with old behaviour that's somewhat broken with multiple interpreters.

The refleaks are one-time, right? Or can you leak an unlimited amount of memory by re-loading the module?

**Comment by picnixz** [comment]
> The refleaks are one-time, right? Or can you leak an unlimited amount of memory by re-loading the module?

Mmh I don't think so actually (or at least I didn't manage to).

**Comment by picnixz** [comment]
> ideally: maintain process-global state, with all Python-level objects being “proxies” to that (i.e. module-level state)

I have a module's state but I'm not sure this is what you have in mind. 

> But neither had priority, so it's left with old behaviour that's somewhat broken with multiple interpreters.

Would `{Py_mod_multiple_interpreters, Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED}` be sufficient to indicate that it's not supported with multiple interpreters?

> I guess it's not very useful outside the main interpreter, so it was left behind.

I'm not sure if this could matter, but for extensions that could be using stdlib modules, wouldn't it be better if they could detect whether they have refleaks by running `python -X showrefcount`? that way if there is a reference leak then the extension could assume that it's likely one their side and not on CPython's side (we could still have refleaks on our side but it's maybe better to reduce them as much as possible).

**Comment by encukou** [comment]
> I have a module's state but I'm not sure this is what you have in mind.

Sorry for the truncated message! I meant *global state is not stored in module-level state*. (edited above)

> Would `{Py_mod_multiple_interpreters, Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED}` be sufficient to indicate that it's not supported with multiple interpreters?

Yes. 
I'd prefer the [global flag](https://docs.python.org/3/howto/isolating-extensions.html#opt-out-limiting-to-one-module-object-per-process), which means you can load it in *any* interpreter, not just the main one, but `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED` is fine.


> wouldn't it be better if [extensions] could detect whether they have refleaks by running `python -X showrefcount`?

Yes. I'm not talking about whether this is a valid issue, but about priorities :)

File: Modules/_cursesmodule.c [hunk 3]
```diff
@@ -199,6 +222,33 @@ static char *screen_encoding = NULL;
...
+        if (!(STATUS)) {                \
+            goto abort;                 \
+        }                               \
+    } while (0)
+
+#define ADD_INT_CONSTANT_OR_ABORT(MODULE, STRING, VALUE)            \
+    do {                                                            \
+        int rc = PyModule_AddIntConstant(MODULE, STRING, VALUE);    \
+        CHECK_RET_CODE_OR_ABORT(rc);                                \
+    } while (0)
```
Comment (vstinner) [inline Modules/_cursesmodule.c:250]: You should move these macros just before their first use, I don't think that it's useful to have them at the top.

File: Modules/_cursesmodule.c [hunk 4]
```diff
@@ -3361,10 +3397,15 @@ _curses_initscr_impl(PyObject *module)
...
     SetDictInt("LINES", LINES);
     SetDictInt("COLS", COLS);
+#undef SetDictInt
 
-    winobj = (PyCursesWindowObject *)PyCursesWindow_New(win, NULL);
+    _cursesmodule_state *state = get_cursesmodule_state(module);
+    PyCursesWindowObject *winobj = (PyCursesWindowObject *)PyCursesWindow_New(state, win, NULL);
+    CHECK_NOT_NULL_OR_ABORT(winobj);
     screen_encoding = winobj->encoding;
     return (PyObject *)winobj;
```
Comment (vstinner) [inline Modules/_cursesmodule.c:3406]: I suggest adding an empty line (after return, before the label) for readability.

**Review by vstinner — COMMENTED** [review]
This PR does "everything" at once, so it's hard to review. I would prefer to split the PR into sub-steps:

* Cleanup the code, prepare PyInit__curses()
* Add a module state
* Convert static types to heap types

File: Modules/_cursesmodule.c (hunk 3 above)
Comment (picnixz) [reply Modules/_cursesmodule.c:250]: Actually, I put them in the "/* Utility Macros */" section where the other macros are being declared.

**Comment by picnixz** [comment]
> I meant global state is not stored in module-level state

What would the the global state consists of? Currently, what's being stored is the type of curses window (`PyCursesWindow_Type`) and the exception type (`PyCursesError`). The curses window type is also exported to the `PyCurses_API` and there is a global variable for `PyCursesError` because some functions need it but do not have access to the module's state.

Those functions are exported in the `PyCurses_API` and test whether curses has been initialized or not. I'm not sure whether I did something wrong because I'm wondering now whether it's possible for them to outlive the module's state, in which case calling the following macro could fail:

```C
#define PyCursesSetupTermCalled                                         \
    if (initialised_setupterm != TRUE) {                                \
        PyErr_SetString(PyCursesError,                                  \
                        "must call (at least) setupterm() first");      \
        return 0; }
```

So maybe we need to live with those refleaks?

**Comment by vstinner** [comment]
> Do you want me to split it into multiple PRs or multiple commits?

Multiple PRs. Start by just creating one, others will follow once the first is merged.

**Comment by picnixz** [comment]
Closing this one since I'll split it in different PRs (and to reduce the number of opened PRs)

**Pull request closed without merging**